### PR TITLE
Add regression test for P3.

### PR DESCRIPTION
**Pull request by ambrad**

### CODE REVIEW

File: components/scream/p3/micro_p3_iso_c.f90 [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+       ok = array_io_write(revap_filename, c_loc(revap_table), Size(revap_table))
+       ok = array_io_write(vn_filename, c_loc(vn_table), Size(vn_table))
+       ok = array_io_write(vm_filename, c_loc(vm_table), Size(vm_table))
+       if (.not. ok) then
+          print *, 'micro_p3_iso_c::p3_init: Error when writing table files.'
+          info = -1
+       end if
+    end if
+    
+    info = 0
```
Comment (ambrad) [inline components/scream/p3/micro_p3_iso_c.f90]: I need to clean up info here so the right value gets returned.

File: components/scream/p3/p3.hpp [hunk 2]
```diff
@@ -1,12 +1,42 @@
...
+#include "scream_kokkos.hpp"
+#include "scream_types.hpp"
 
-int p3_stub();
+namespace scream {
+namespace p3 {
 
-}  // namespace P3
-}  // namespace Scream
+template <typename exespace = Kokkos::DefaultExecutionSpace,
```
Comment (ambrad) [inline components/scream/p3/p3.hpp]: Remove this generality and rename to FortranData. The C++ data should group all arrays in a column together so the data are close; hence, multi-D arrays over (col,lev) won't be used. Moreover, the lev dim should be compile time. In summary, this current Data should be for Fortran calling only.

File: components/scream/p3/p3_impl.hpp [hunk 3]
```diff
@@ -0,0 +1,40 @@
+#include "p3.hpp"
```
Comment (ambrad) [inline components/scream/p3/p3_impl.hpp]: Move this to p3.cpp when Data is converted to FortranData and stripped of its template params.

File: components/scream/util/scream_catch_main.cpp [hunk 4]
```diff
@@ -0,0 +1,16 @@
+#define CATCH_CONFIG_RUNNER
+
+#include "catch2/catch.hpp"
+
+#include "scream_kokkos.hpp"
+#include "scream_util.hpp"
+
+int main (int argc, char **argv) {
+  scream::util::activate_floating_point_exceptions_if_enabled();
```
Comment (ambrad) [inline components/scream/util/scream_catch_main.cpp]: Make a scream_initialize and scream_finalize. Move the Kokkos, as well as the FPE, calls into it.

**Comment by ambrad** [comment]
@PeterCaldwell This PR is ready for review.

**Comment by ambrad** [comment]
By the way, we'll likely never get Codacy to give a 100% pass. Code coverage is < 100% whenever there are unexercised exceptions, and we definitely don't intend to exercise all of our exceptions. The quality review makes mistakes; e.g., it currently thinks there is an unused struct member, but that member is definitely used. Thus, we'll have to be content with failing Codacy checks and just use its data when useful and otherwise ignore it.

**Comment by PeterCaldwell** [comment]
Ok, thanks for the Codacy explanation. I was wondering.

**Comment by jgfouca** [comment]
@ambrad , bummer. Is there any way to give Codacy hints to ignore certain things? Pylint supports that through specially formatted code comments.

**Comment by PeterCaldwell** [comment]
@jgfouca - I'm curious what you did before approving this PR. My thought was "wow, there's a lot of code changes here - this will take me a while to get through"... My current plan is to understand the initialization changes and to exercise the test before approving.

**Comment by PeterCaldwell** [comment]
@jgfouca - thanks for the explanation. I think I'll stick to my plan for review, which will probably take me until Monday to do. @ambrad - thanks for pointing out the README. I'll check it out and make corrections as needed.

**Comment by PeterCaldwell** [comment]
When I try to cmake kokkos on LLNL or NERSC machines, I get the errors below. My interpretation is that kokkos assumes ${CMAKE}/core/src/Makefile.generate_header_lists and Makefile.generate_build_files files exist, but they don't... Thoughts?

/g/g11/caldwep/gitwork/kokkos/cmake/Makefile.generate_cmake_settings:7: /usr/tce/packages/cmake/cmake-3.5.2/bin/cmake/core/src/Makefile.generate_header_lists: Not a directory
/g/g11/caldwep/gitwork/kokkos/cmake/Makefile.generate_cmake_settings:8: /usr/tce/packages/cmake/cmake-3.5.2/bin/cmake/core/src/Makefile.generate_build_files: Not a directory
make: *** No rule to make target `/usr/tce/packages/cmake/cmake-3.5.2/bin/cmake/core/src/Makefile.generate_build_files'.  Stop.
CMake Error at CMakeLists.txt:43 (message):
  Kokkos settings generation failed:

  env;KOKKOS_SRC_PATH=/g/g11/caldwep/gitwork/kokkos;KOKKOS_PATH=/g/g11/caldwep/gitwork/kokkos;KOKKOS_INSTALL_PATH=/p/lscratchh/caldwep/scream/kokkos;KOKKOS_ARCH=None;KOKKOS_DEVICES=OpenMP,Serial;KOKKOS_DEBUG=yes;KOKKOS_OPTIONS=disable_profiling,disable_deprecated_code,disable_dualview_modify_check;KOKKOS_USE_TPLS=librt
  make -f /g/g11/caldwep/gitwork/kokkos/cmake/Makefile.generate_cmake_settings
  CXX=/usr/tce/bin/c++ generate_build_settings

**Comment by ambrad** [comment]
@PeterCaldwell Thanks for the error report. I've modified the Kokkos build instructions in README.md; I believe it should resolve the problem.

IMPORTANT: Rm the entire kokkos repo and clone from scratch. The failed cmake run has polluted it with cmake cache junk that will mess things up. In the modified instructions, I build out of source (as one should, anyway; oversight on my part not to do that in the first place) and install out of source, so we won't pollute the Kokkos repo anymore.

**Comment by PeterCaldwell** [comment]
With @ambrad 's README mods I was able to compile and run the baseline and comparison tests. I found the directory structure for building/running to be confusing, so I modified README.md to be more explicit about this. I'm still unclear where data files created by running baselines are stored and (more importantly) how I should compare 2 versions of the code. Is the idea to rebuild scream without re-baselining, then to do a comparison with the old baseline? This might be worth articulating in the README.

**Comment by ambrad** [comment]
Baseline and table files for tests are in p3/tests. I'll update the readme re: how to use the baseline file.

**Comment by ambrad** [comment]
If there is interest in explicitly seeing generated data files (e.g., baseline and memoized table files), we could consider a cmake variable CMAKE_BUILD_DATA_DIR with a default of "data" w/in the build. But a user could point it anywhere.

**Comment by ambrad** [comment]
By "new P3 implementation" I mean the Kokkos one, for example.

**Comment by PeterCaldwell** [comment]
About knowing where data files reside - I'm mildly interested in this but I wouldn't make it a high-priority.

**Comment by PeterCaldwell** [comment]
about "new P3 implementation": so if I understand you correctly, you mean a new implementation that you intend to keep as an option alongside the existing implementation as opposed to just another incarnation of the same code after fixing a bug (for example)?

**Comment by ambrad** [comment]
Yes. In more detail: I think of the "existing impl" as the "reference impl". It should give correct answers. Rebaselining will be necessary when a change is made b/c a bug is fixed in the ref impl. A new impl is a performance-oriented one that should give the same answer as the ref impl but faster and on all architectures.

**Comment by ambrad** [comment]
By the way, eventually we can make the Kokkos impl the reference and stop referring to the Fortran.

**Comment by PeterCaldwell** [comment]
It seems odd to me that README.md is in the scream/components/scream/ directory since it is focused only on building/running tests. Would it make more sense to put it under scream/components/scream/test/? or to call it README_TESTS.md?

**Comment by ambrad** [comment]
I'm ok with placing it anywhere or calling it anything, so please feel free to move or rename it. However, my opinion is that a typical user will want a README that tells them how to do the basic things the scream component can do, such as build as a standalone correctness and performance testing framework. Since I worked on the regression test, I put that material in the README. But when folks do other things, they should include a section in the README on it. One did not previously exist, so I created it. A different opinion might be that README.md should be short and say what the scream component is, and that's it. Then README_TESTING.md, e.g., or maybe README_STANDALONE.md, would contain the material generated in this PR. Again, any of these or something else is fine with me.

**Comment by bartgol** [comment]
I think a README.md file is ok. I am not fond of options of the form README_XYZ.md. I think README.md should contain info about 1) the package, 2) how to build the standalone scream. I also think it _could_ be useful if the test directory also contained a README file. In that case, the main README.md could simply refer to that file for more info.

**Comment by ambrad** [comment]
Just for clarity, I want to point out that the README currently exists at exactly the level the user needs to know about to build the standalone scream component(s), i.e., at the level to which you point your cmake command.

**Comment by PeterCaldwell** [comment]
Ok, your MAKEFILE.md explanation makes sense. Let's just add additional sections to MAKEFILE.md as we implement new functionality. Thanks for the explanation.

**Pull request merged**